src/main/java/com/atlassian/jira/cloud/jenkins/listeners/AutoBuildsListener.java

```diff
@@ -0,0 +1,261 @@
+package com.atlassian.jira.cloud.jenkins.listeners;
+
+import com.atlassian.jira.cloud.jenkins.buildinfo.service.MultibranchBuildInfoRequest;
+import com.atlassian.jira.cloud.jenkins.common.factory.JiraSenderFactory;
+import com.atlassian.jira.cloud.jenkins.common.response.JiraSendInfoResponse;
+import com.atlassian.jira.cloud.jenkins.common.service.IssueKeyExtractor;
+import com.atlassian.jira.cloud.jenkins.deploymentinfo.client.model.State;
+import com.atlassian.jira.cloud.jenkins.deploymentinfo.service.ChangeLogIssueKeyExtractor;
+import com.atlassian.jira.cloud.jenkins.util.BranchNameIssueKeyExtractor;
+import com.atlassian.jira.cloud.jenkins.util.JenkinsToJiraStatus;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepEndNode;
+import org.jenkinsci.plugins.workflow.cps.nodes.StepStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+import java.util.stream.Collectors;
+
+/**
+ * This class should listen to the events from Jenkins pipeline and send 2 builds: one in
+ * "IN_PROGRESS" state and the other one in the "final" or "resulting" state.
+ *
+ * <p>Challenges to keep in mind:
+ *
+ * <p>- order of the events: we shouldn't send IN_PROGRESS event if the final result has already
+ * been sent
+ *
+ * <p>- uncertainty about issue keys: they appear at some point during the execution of the pipeline
+ * and might not be in place at the moment we need them.
+ */
+public class AutoBuildsListener implements SinglePipelineListener {
```

Comment:
This is the entry point to the new "send builds automatically" feature. The listener is notified on each node in a pipeline and decides when to send a build event to Jira.